sentry_sdk/utils.py

```diff
@@ -243,9 +243,22 @@ def datetime_from_isoformat(value):
     # type: (str) -> datetime
     try:
         return datetime.fromisoformat(value)
-    except AttributeError:
+    except (AttributeError, ValueError):
         # py 3.6
-        return datetime.strptime(value, "%Y-%m-%dT%H:%M:%S.%f")
+        timestamp_format = (
+            "%Y-%m-%dT%H:%M:%S.%f" if "." in value else "%Y-%m-%dT%H:%M:%S"
+        )
+        if value.endswith("Z"):
+            value = value[:-1]
+        else:
+            plus_idx = value.rfind("+")
+            if plus_idx != -1:
+                tz_colon_index = value.rfind(":", plus_idx)
+                if tz_colon_index != -1:
+                    value = value[:tz_colon_index] + value[tz_colon_index + 1 :]
+                timestamp_format += "%z"
+
+        return datetime.strptime(value, timestamp_format).astimezone(timezone.utc)
```

Comment:
I'm not sure if we can assume UTC here. Since folks can provide their own timestamp in `add_breadcrumb` (and it seems like some are using this to provide strings instead of datetimes, so who knows what the timestamps actually look like), the timezone could be anything.

(Can't wait to drop 3.6.)

One thing we could do is take in `datetime.fromisoformat` from 3.7+, then we wouldn't have to do any parsing by hand. We've done this before with [LRU cache](https://github.com/getsentry/sentry-python/blob/master/sentry_sdk/_lru_cache.py).

Reply:
fromisoformat still blows up in py3.7 when there's a colon in timezone part so it is 3.8+ that is safe. 

Regarding casting to UTC, my intention was to _convert_ to UTC, not coerce. Is that not what astimezone does? (vs replace(tzinfo=))

Reply:
> Regarding casting to UTC, my intention was to convert to UTC, not coerce. Is that not what astimezone does? (vs replace(tzinfo=))

I was referring to the parsing part but I misread it originally -- anything ending in `Z` or `+xx:xx` is working great, it's just the `-yy:yy` case that's not handled since we don't add the `%z` to the format in that case

Reply:
Yup, great catch. Completely forgot about the `-` thing. Update coming.